avoid collect/copy of payload after encode

akonring opened the pull request:
Closes: EspressoSystems/jellyfish#417 and #2115
<!-- See here for keywords in Github -->
<!-- https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword -->

### This PR: 
Addresses the issue of collect/copy of the encoded payload like here:
https://github.com/EspressoSystems/HotShot/blob/15c776d32001ce88edf2a87f3c520301da22a582/crates/hotshot/src/lib.rs#L217-L226
This is done by changing the output type of `BlockPayload::encode` to `Result<Arc<[u8]>, Self::Error>` and thereby avoiding collect/copy when using it in fx `vid_commitment`.

<!-- Describe what this PR adds to HotSHot -->
<!-- E.g. -->
<!-- * Implements feature 1 -->
<!-- * Implements feature 2 -->
<!-- * Fixes bug 3 -->

**NB: This has implications for components downstream (espresso-sequencer, hotshot-query-service, hotshot-builder-core) using this trait interface.**

### This PR does not: 
<!-- Describe what is out of scope for this PR, if applicable.  Leave this section blank if it's not applicable -->
<!-- * Implement feature 3 because that feature is blocked by Issue 4   -->

### Key places to review: 
<!-- Describe key places for reviewers to pay close attention to -->
<!-- * file.rs, `add_integers` function -->

<!-- ### How to test this PR:  -->
<!-- Optional, uncomment the above line if this is relevant to your PR -->
<!-- If your PR can be tested through CI there is no need to add this section -->
<!-- * E.g. `just async_std test` -->

<!-- Complete the following items before creating this PR
* Are the proper people tagged to review it?
* Have you linked an issue to this PR?   -->

ss-es commented on crates/example-types/src/block_types.rs:
```diff
@@ -150,7 +149,7 @@ impl BlockPayload for TestBlockPayload {
     }
 
-    fn encode(&self) -> Result<Self::Encode<'_>, Self::Error> {
+    fn encode(&self) -> Result<impl AsRef<[u8]> + Send, Self::Error> {
```
Looks fine to me overall, but just wondering -- is there any reason we need the more general `impl AsRef<[u8]> + Send` rather than just e.g. `Arc<[u8]>`?

ggutoski replied:
Not really. Some of us have kinda started using `AsRef<[u8]>` instead of `&[u8]` just to maximize generality. I don't have a strong opinion here and could be convinced to do something else, including just `&[u8]`.

ss-es replied:
I don't have a strong opinion either.

Is it fine if the result of `encode` gets cast to a `Vec<u8>`? Would this still avoid the collect/copy? If so, I think the `impl` is okay.

ggutoski replied:
> Is it fine if the result of `encode` gets cast to a `Vec<u8>`? Would this still avoid the collect/copy?

Uh no. The only way to return a `Vec<u8>` without copying data is to consume self.

ss-es replied:
Then I think this should return a concrete non-Vec type, because e.g. in https://github.com/EspressoSystems/HotShot/pull/2964/files#diff-afe45f19786400fa4404b4255749941ded407b70ba5a1f26a3e12e70dfc76e08R216-R222 I believe the result gets cast to Vec<u8>

ggutoski commented on crates/hotshot/src/lib.rs:
```diff
@@ -213,16 +213,17 @@ impl<TYPES: NodeType, I: NodeImplementation<TYPES>> SystemContext<TYPES, I> {
-            saved_payloads.insert(anchored_leaf.get_view_number(), encoded_txns);
+            saved_payloads.insert(
+                anchored_leaf.get_view_number(),
+                encoded_txns.as_ref().to_vec(),
```
omg I wasn't even aware that we're copying the payload here anyway. Oh well, it's good practice to clone it only when it's actually needed

ggutoski commented on crates/task-impls/src/transactions.rs:
```diff
@@ -126,15 +126,16 @@ impl<
                     broadcast_event(
                         Arc::new(HotShotEvent::BlockRecv(
-                            encoded_transactions,
+                            encoded_transactions.as_ref().to_vec(),
```
ugh. I suspect the compiler is smart enough not to copy the entire payload here, since `block_data` goes out of scope in this block anyway. Still this is bad practice.

@bfish713 lots of events seem to demand an owned block payload. Is that necessary? If so then we should be passing ownership around instead of copying (eg. via `to_vec()`) whenever possible.

I wonder whether we should make a `BlockPayload` method that consumes self and returns the owned payload (and other stuff, such as the namespace table) for use cases such as this. 🤔 

Or... I suspect most events could just take ownership of the `BlockPayload` directly instead of extracting it's binary payload. It's certainly true for this `BlockRecv` event.

ggutoski reviewed (commented):
Looks pretty good. But since we're changing the API I think it's best to take the time to resolve some questions. To summarize comments below:
- Should we switch convention away from `AsRef<T>`?
- Can we avoid explicit copying into event data? eg. Events take a `BlockPayload`, or we add a `BlockPayload` method to consume self and relinquish ownership of the underlying data.

akonring commented:
Really appreciate the comments. 

The simplicity of the `Arc<[u8]>`-based approach is appealing. I've updated the PR so that it uses this.
Note, this is merely to have a concrete suggestion to analyze trade-offs. In other words, I'm happy to revert back or go in another direction if this seems more appealing to you.

Also, did not implement the updated (Arc-based) trait interface on the espresso-sequencer side, yet. So still unknown if this interface design will work seamlessly downstream.

akonring commented on crates/types/src/data.rs:
```diff
@@ -457,7 +456,7 @@ impl<TYPES: NodeType> Leaf<TYPES> {
             upgrade_certificate: None,
             block_header: block_header.clone(),
-            block_payload: Some(payload),
+            block_payload: Some(payload.clone()),
```
Good catch. Fixing this now.

dailinsubjam replied:
I'm surprised that `lint` didn't catch this.

ggutoski commented on crates/task-impls/src/helpers.rs:
```diff
@@ -44,13 +44,13 @@ pub async fn broadcast_event<E: Clone + std::fmt::Debug>(event: E, sender: &Send
 #[allow(clippy::panic)]
 pub async fn calculate_vid_disperse<TYPES: NodeType>(
-    txns: Vec<u8>,
+    txns: Arc<Vec<u8>>,
```
```suggestion
    txns: &[u8],
```
Should work with just `&[u8]` no? I don't think it should ever have been a `Vec<u8>`. That was forcing a clone for each call to the function. 😬 It's a good thing we're cleaning this up.

akonring replied:
Yes, this is tricky with lifetimes because `txns` is passed into `spawn_blocking`.

Maybe there is another more elegant approach but keeping this as an `Arc<[u8]>` and just handing out a reference seems to work.

ggutoski commented on crates/example-types/src/block_types.rs:
```diff
@@ -29,7 +30,7 @@ impl TestTransaction {
     /// # Errors
     /// If the transaction length conversion fails.
-    pub fn encode(transactions: Vec<Self>) -> Result<Vec<u8>, BlockError> {
+    pub fn encode(transactions: &Vec<Self>) -> Result<Arc<Vec<u8>>, BlockError> {
```
```suggestion
    pub fn encode(transactions: &[Self]) -> Result<Vec<u8>, BlockError> {
```
- `transactions` type should be either `&[Self]` or `Vec<Self>` but not `&Vec<Self>`
- No need to return `Arc<Vec<u8>>` because the returned `Vec` is newly created. The caller can wrap it in `Arc` if needed.

akonring replied:
Ok. Modified according to your suggestion so the caller creates the `Arc` (if needed).

ggutoski reviewed (changes requested):
Looking good. A few suggestions in the comments. Some generic notes here:
- Use `Arc` as rarely as possible. If you can remove an `Arc` and it still compiles then that's good. `Arc` should only be used when you need to clone: if you're calling `x.clone` then `x` should be an `Arc`, otherwise `x` should not be an `Arc`.
- I'm surprised you need `.as_ref()` everywhere. Can't you use `&` or nothing?
- Question on code convention: Shall we use `Arc::clone()` instead of `.clone()`? More details in a comment; I didn't flag all the places where this occurs.

ggutoski commented:
> However, I'm a little worried about the performance hit from the guarantee of atomic operations in Arc versus normal memory access.

I don't know the answer to that question. But presumably cloning a vec requires some form of thread safety at some level just to allocate the memory no?

We could try with `Rc` instead of `Arc` and see whether the compiler yells. 😛

akonring commented on crates/types/src/traits/block_contents.rs:
```diff
@@ -119,10 +111,11 @@ pub trait TestableBlock: BlockPayload + Debug {
 #[allow(clippy::panic)]
 pub fn vid_commitment(
-    encoded_transactions: &Vec<u8>,
+    encoded_transactions: &Arc<Vec<u8>>,
```
Yes, agree. Should be fixed now.

akonring commented on crates/types/src/traits/block_contents.rs:
```diff
@@ -131,13 +124,14 @@ pub fn vid_commitment(
 #[allow(clippy::panic)]
 pub fn precompute_vid_commitment(
-    encoded_transactions: &Vec<u8>,
+    encoded_transactions: &Arc<Vec<u8>>,
```
same

akonring commented on crates/task-impls/src/consensus/mod.rs:
```diff
@@ -672,7 +672,7 @@ impl<TYPES: NodeType, I: NodeImplementation<TYPES>, A: ConsensusApi<TYPES, I> +
                                 {
                                     let payload = BlockPayload::from_bytes(
-                                        encoded_txns.clone().into_iter(),
+                                        encoded_txns.clone(),
```
Think we reached consensus on this. Changed to `Arc::clone`.

akonring commented on crates/task-impls/src/da.rs:
```diff
@@ -132,7 +132,8 @@ impl<TYPES: NodeType, I: NodeImplementation<TYPES>, A: ConsensusApi<TYPES, I> +
 
-                let encoded_transactions_hash = Sha256::digest(&proposal.data.encoded_transactions);
+                let encoded_transactions_hash =
+                    Sha256::digest(proposal.data.encoded_transactions.as_ref());
```
Yes, `as_ref()` was necessary when the type of `proposal.data.encoded_transactions` was `Arc<Vec<u8>>`. This was due to Vec not implementing `AsRef<[u8]>` and thus no auto-coercion without the `.as_ref()`. 

Just reverted back to `Arc<[u8]>` so should be fixed according to your suggestion now.

akonring commented:
Modified according to the comments above. Specifically:
1. Changed to idiomatic use of `Arc::clone`.
2. Limit the use of `Arc` for payload values to a minimum.
3. Changed back to `Arc<[u8]>` instead of `Arc<Vec<u8>>` since I did not identify any advantage in using `Vec` and we can save a memory allocation for vec capacity.

jparr721 reviewed (approved):
This is awesome

shenkeyao reviewed (dismissed):
Thanks for doing this. Looks good!

ggutoski commented on crates/example-types/src/block_types.rs:
```diff
@@ -150,8 +146,9 @@ impl BlockPayload for TestBlockPayload {
-    fn encode(&self) -> Result<Self::Encode<'_>, Self::Error> {
-        Ok(TestTransaction::encode(self.transactions.clone())?.into_iter())
+    fn encode(&self) -> Result<Arc<[u8]>, Self::Error> {
+        let encoded = TestTransaction::encode(&self.transactions);
+        encoded.map(Arc::from)
```
```suggestion
        TestTransaction::encode(&self.transactions).map(Arc::from)
```
code golf improvement? 😉

ggutoski reviewed (dismissed):
Looks great! A couple tiny questions below, plus:

- As discussed offline, we should add a lint for `Arc::clone` as demo'ed in #3007. I suppose the best plan is to merge this PR then do a separate PR for that lint.

Approved. Tweak and merge at your discretion.

akonring commented on crates/task-impls/src/helpers.rs:
```diff
@@ -44,7 +44,7 @@ pub async fn broadcast_event<E: Clone + std::fmt::Debug>(event: E, sender: &Send
 #[allow(clippy::panic)]
 pub async fn calculate_vid_disperse<TYPES: NodeType>(
-    txns: Vec<u8>,
+    txns: Arc<[u8]>,
```
same as above:
> Yes, this is tricky with lifetimes because txns is passed into spawn_blocking. Maybe there is another more elegant approach but keeping this as an Arc<[u8]> and just handing out a reference seems to work.
